src/Compilers/Core/Portable/Syntax/ChildSyntaxList.Segment.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public partial struct ChildSyntaxList
+    {
+        internal readonly struct Segment
+        {
+            private readonly SyntaxNode? _node;
+            private readonly int _count;
+            private readonly int _initialIndex;
+
+            internal Segment(SyntaxNode node, int initialIndex)
+            {
+                _node = node;
+                _count = CountNodes(node.Green);
+                _initialIndex = initialIndex;
+            }
+
+            internal Segment(SyntaxNode node, int initialIndex, int terminalIndex)
+            {
+                _node = node;
+                _count = terminalIndex;
+                _initialIndex = initialIndex;
+            }
+
+            public Enumerator GetEnumerator()
+            {
+                Debug.Assert(_node is object);
```

Comment:
why is this assert here, and not in the constructor?

Reply:
The first constructor expects a non-null node and operates with that assumption,
the second constructor supports a null node when initialIndex >= terminalIndex although that scenario is an edge-case...

In `GetEnumerator` since a null node is supported under the same limited circumstances,
perhaps a null suppression (i.e. '!') would be more appropriate here...

Reply:
> the second constructor supports a null node when initialIndex >= terminalIndex although that scenario is an edge-case...

I don't understand when this happens.  Do you have a test that exercises that case?

regardless, if this can happen, it needs a doc explaining things.

Reply:
`default(Segment)` and `new Segment(null!, 0, 0)` are the main instances in which this supported edge case would happen... [relevant comment](https://github.com/dotnet/roslyn/pull/69740/files/bef17ddf5a0fdb96513596e545b5969e22389420#r1332345034)

Reply:
Ok.  Please provide tests of GetPrevious/NextToken on real trees that then hit these cases.  And doc in the code that tehse are the specific cases this is handling.